linera-views/src/journaling.rs

```diff
@@ -231,28 +250,62 @@ where
+    /// batch, using one write-operation per block. The last transaction also updates the
+    /// journal header with the final number of blocks. As a result, the constraints of
+    /// the underlying database are respected when a transaction is written if the
+    /// following conditions are met:
+    ///
+    /// (1) the number of blocks per transaction doesn't exceed `K::MAX_BATCH_SIZE - 1`;
+    ///
+    /// (2) the total size of BCS-serialized blocks in a transaction doesn't exceed
+    /// `K::MAX_BATCH_TOTAL_SIZE - (K::MAX_BATCH_SIZE - 1) * sizeof(block_key) - sizeof(header_key) - sizeof(bcs_header)`
+    ///
+    /// (3) the size of each BCS-serialized block doesn't exceed `K::MAX_VALUE_SIZE`.
+    ///
+    /// (4) [static invariant] `block_key` and `header_key` don't exceed `K::MAX_KEY_SIZE` and
+    /// `bcs_header` doesn't exceed `K::MAX_VALUE_SIZE`.
+    ///
+    /// NOTE:
+    /// * Since a block must contain at least one operation and M bytes of the
+    /// serialization overhead (typically M = 2 or 3 bytes of vector sizes), condition (3)
+    /// requires that each operation in the original batch satisfies:
+    ///     `sizeof(key) + sizeof(value) + M <= K::MAX_VALUE_SIZE`
+    ///
+    /// * Similarly, a transaction must contain at least one block so it is desirable that
+    /// the maximum size of a block insertion `1 + sizeof(block_key) + K::MAX_VALUE_SIZE`
+    /// plus M bytes of overhead doesn't exceed the threshold of condition (2).
     async fn write_journal(
         &self,
-        simplified_batch: K::Batch,
+        batch: K::Batch,
         base_key: &[u8],
     ) -> Result<JournalHeader, K::Error> {
-        let mut iter = simplified_batch.into_iter();
+        let mut iter = batch.into_iter();
         let mut block_batch = K::Batch::default();
         let mut block_size = 0;
         let mut block_count = 0;
         let mut transaction_batch = K::Batch::default();
         let mut transaction_size = 0;
         while iter.write_next_value(&mut block_batch, &mut block_size)? {
             let (block_flush, transaction_flush) = {
-                if iter.is_empty() || block_batch.len() == K::MAX_BATCH_SIZE - 2 {
+                if iter.is_empty() || transaction_batch.len() == K::MAX_BATCH_SIZE - 1 {
                     (true, true)
                 } else {
                     let next_block_size =
                         block_size + iter.next_value_size()? + block_batch.overhead_size();
                     let next_transaction_size = transaction_size + next_block_size;
                     let transaction_flush = next_transaction_size > K::MAX_BATCH_TOTAL_SIZE;
-                    let block_flush = transaction_flush || next_block_size > K::MAX_VALUE_SIZE;
+                    let block_flush = transaction_flush
+                        || block_batch.len() == K::MAX_BATCH_SIZE - 2
+                        || next_block_size > K::MAX_VALUE_SIZE;
```

Comment:
I'm inclined to blame this existing inequality (rather than my change line 299) for the failed test.

Reply:
I am also more inclined to think the correction is needed, but I will do more extensive checks.